### ISSUE TITLE
Kotlin Support

### ISSUE THREAD

**Issue by aforcier** [wordpress-mobile/WordPress-FluxC-Android#547]
This issue tracks progress of adding Kotlin support for FluxC. There are a few blocking items to address before we can add it:

### 1. Payload `error` field overwriting

Kotlin is not happy with our overwriting the `error` field in payloads to use custom errors (see, e.g., [AuthenticateErrorPayload](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/3c2dc961e39b2aba7fe98b37597b640fa7d78203/fluxc/src/main/java/org/wordpress/android/fluxc/store/AccountStore.java#L57-L65)). Any Kotlin code we wrote that needed to access a payload with an overwritten type would cause a compiler error. This is because the underlying `BaseNetworkError` is exposed to Kotlin through a `getError()` method, so there's a conflict between the base `BaseNetworkError getError()` and the overwriting, custom error.

I think the least painful way to solve this is to stop overwriting the error, and instead let `Payload` accept a generic parameter for the `error` type. We won't be able to default to `BaseNetworkError` anymore (we didn't want other errors to inherit from this), so any `Payload` subclasses not using a custom error type will need to use `extends Payload<BaseNetworkError>`. This is a bit flimsy since there's nothing forcing `Payload` subclasses to parameterize `Payload`, giving those subclasses a useless `error` type. So, we'll enforce parameterizing `Payload` subclasses using a Checkstyle rule.

I'm confident we can find a better fix in the future, especially if we move the `Payload` and its subclasses to Kotlin - but for now this is the most painless way to unblock `Payload` usage in Kotlin code.

#### Tasks:
* [x] Generify `Payload`, and enforce its parameterization when subclassed (PR: #548)

### 2. Linting and Checkstyle

Kotlin for Android is still quite new, and we don't have the same range of options available as we do for Java. [`ktlint`](https://github.com/shyiko/ktlint) looks like a great option, but there are a few problems:

<ol>
<li>No IDEA plugin

Checkstyle's IDEA plugin allows checkstyle violations to be highlighted in the IDE. There's no such plugin yet for `ktlint`, which means we're likely not to notice violations until the commits have been pushed and Travis has run `ktlint` on the commandline and failed the build.

We can reflect some of `ktlint`'s rules in the Android Studio settings, but we can't do it for everything. Using a pre-commit hook for ktlint would probably be a good compromise here.
</li>

<li>Maintaining checkstyle rules in `*.kt` files

I'm looking into either moving over our checkstyle rules (in particular, the regex ones preventing things like 'dotcom') to `ktlint`, or configuring checkstyle to run against `.kt` files (and seeing whether we get some false positives there due to the Java rules). My first pass on this suggests we can run our checkstyle against both, but some more testing is needed.
</li>

<li>Sharing configs

@maxme has already done the core work needed for this in #539. We can add Kotlin-specific rules to the shared IDEA configs to help ensure consistency, and use the IDE as much as possible for style/lint error flagging since there's no plugin for `ktlint`.
</li>
</ol>

#### Tasks: (all done in #566)
* [x] Enable Kotlin
* [x] Add `ktlint`, and run everywhere (FluxC source, tests, example apps)
* [x] Add `ktlintFormat`
* [x] Sort out checkstyle support
* [x] Sort out which configs can/should be shared

### 3. `Action#name()` conflict

Kotlin [does not like](https://youtrack.jetbrains.com/issue/KT-14115) that we extend `IAction` (which is [an interface with a `String name()` method](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/9e535369eb7dfef24c844e90443b3c634d3c4bde/fluxc-annotations/src/main/java/org/wordpress/android/fluxc/annotations/action/IAction.java#L4)) in `enum`s, which have their own `name()` method. I think we can safely drop `name()` from `IAction` and resolve this, but some testing is needed.

#### Tasks:
* [x] Resolve `name()` conflict so we can add new `Action` classes in Kotlin (PR: #557)

**Comment by aforcier** [comment]
All remaining items for this are addressed in #566.

**Comment by aforcier** [comment]
Added in #566 🎉
